### DISCUSSION TITLE
ANSI Support

### DISCUSSION THREAD

**Discussion by ghost** [frnno967/maxiterm#5] [Show and tell]
Hi

I added ANSI colour support / cursor control, and support for the arrow keys / page up / page down / home and end keys in my fork.

I move xmodem up loading and downloading to ALT+PAGEUP and ALT+PAGEDOWN, so page up and page down are sent through the terminal.

I'm still experimenting/testing, but if you'd like to see it's here: https://github.com/apamment/maxiterm

If you want I can do a PR, or if you want to pick parts out of my fork and put them in yours feel free :)

Andrew

**Comment by frnno967** [comment]
Hello Andrew.

Let me start by saying that for lack of better adjectives, your work is AMAZING. I can't believe you were able to graft in some ANSI support so fast and it works so well already! Definitely will be a permanent addition to the program. I couldn't believe my eyes when I saw it work this morning, it nearly made me cry it looked so amazing.

There are still some bugs. I was only able to run one test before I left for work but my Input handling routine stopped working after I logged off of a BBS, meaning I couldn't type anything and Maxiterm wasn't responding to any ALT keys. I had to CONTROL-C to get out of the program. And while I was connected to a BBS, there were a couple of instances where data was dropped/not displayed but I think that's because I need to change the way data is received for display. Right now that routine runs when called by a COM interrupt but that code needs to be moved into the main loop of the program and take advantage of the buffering available on the CMM2. I made an alpha version of the program with that type of arrangement and it worked very well and much faster than it does now, but I had trouble with uploading/downloading breaking after that so I shelved it for the moment. 

Would you be willing to collaborate for a while on getting ANSI really cleaned up and working well? I definitely want to pull it into the main release and have you be a collaborator if you're willing.

**Reply by ghost** [reply]
Hmm, it's odd that the input handling routine stopped working, as I haven't seen that - I wonder if it's differences with our "modem"s or differences with the BBS, causing different results?

My arduino modem I made for a Z80 machine running CP/M and it slows down the rate of receiving characters as the Z80 although it used 115200 baud, couldn't keep up and would drop characters - I suspect that slow down is why I'm not seeing the dropped characters.

I'm hoping to get something to run ZiModem on in the mail soon, so we should have similar modems then, is your BBS public? My BBS is located at happylandbbs.com on port 11892 (for telnet).

I am keen to help with getting this working, I am very new to CMM2 and MMBasic though, but I have written a BBS terminal program (for linux / windows), so I know my way around that side of things fairly well.

Would it be possible to see the alpha version of the program you made?

**Reply by ghost** [reply]
I got my ESP32 thing today in the mail, took me a while to figure out I had to hold the button down to flash it.. but it's working well. I did notice an error in the ansi code (switching to bold by itself didn't change the colour to bold if the colour wasn't changed in the same code) that's fixed now I think.

**Reply by frnno967** [reply]
Sounds good. I wasn't able to duplicate the error yet where it locked up the input handling but I did find a few other bugs that are not major. Like when I pull up the Help Menu (ALT-H) or other screens it draws using the last color that your code rendered. Easy to fix by invoking the default color choice before opening the dialogue.

If you email me at recstudio@gmail.com I can send you the alpha version of the code where I moved the rendering into the main loop instead of in the RX serial interrupt section. After reading [this thread](https://www.thebackshed.com/forum/ViewTopic.php?TID=13045&PID=158835#158835) on The Back Shed it made me realize that the way I'm doing the RX is wrong and causes problems. By getting rid of the interrupt subroutine when data is RX it dramatically sped up the rendering but it also completely broke the ALPHA quality Xmodem code that is in Maxiterm. So the way forward is definitely to redesign everything around using the RX buffering and no interrupt, but it's going to break quite a few things so I haven't released anything yet with that design. But happy to send that version to you for your evaluation of it.
